libp2p/protocols/pubsub/rpc/message.nim

```diff
@@ -57,19 +57,22 @@ proc verify*(m: Message, p: PeerInfo): bool =
   else:
     libp2p_pubsub_sig_verify_failure.inc()
 
+# cannot explicit init but:
+# Thread locals are initialized to binary zero (usually default(T)). (Araq)
+var tSeqno {.threadvar.}: uint64 
```

Comment:
this should be a field in `PubSub` - the more threadvars a library uses, the less flexible it is to use - it's also harder to test, has worse platform support, generates poor code and a number of other reasons

Reply:
Agree, this is a pretty arbitrary number either way, in fact it should have probably been left unspecified in the spec, because `linearly increasing` doesn't add anything at all, since the number only makes sense for the peer emitting the message all other peers shouldn't and cannot make any assumptions about it. So, just having a counter in `PubSub` should be enough.

Reply:
I guess this week is full of critics for me...

I honestly don't understand your adversities towards TLS.
It's an amazing tool that works very well in many scenarios. And it's easy to debug and generates fine code too. It's supported everywhere ( I used even in nuttx os ). Gcc and clang support emulation for non conforming platforms too... if you know what you are doing there is absolutely no issue in using it.

That said if this is a nim related issue I can understand.

Reply:
Oh, didn't realize it came across as a criticism, it definitely isn't a knock on you and I have no issues with TLS, @arnetheduck might have his reservations 😄 . But in this case, a TLS would really make no sense, since we already have a "singleton" object, which is the `*sub` instance.

Reply:
Fair enough but guys this is a `init` proc... am I supposed to stick the singleton in it? it's a bit an anti-pattern

Reply:
I didn't mean a singleton in a literal sense, as in the pattern, I mean it as in "we only have one instance of pubsub per switch", which means that we can just stick a counter in it and forget about it (more or less, read my comment here https://github.com/status-im/nim-libp2p/pull/253#issuecomment-657922172)

Reply:
well yeah can just pass the seqno by value to init  😄

Reply:
I would actually put it here somewhere - https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/pubsub/pubsub.nim#L52-L62.

Reply:
> adversities towards TLS

a global, and a thread-local one at that, generally limits the options available to the user of the library - the more "helpful" features you use, the less generally useful the library becomes, because it is no longer a self-contained box but rather has a side effect on the entire application - it becomes harder to run 2 switches, the code is no longer deterministic, you can no longer test it etc etc.

such things are best left for the user of the library to decide - where the switch memory should be allocated, how it should be managed, how it should be isolated as a subsystem etc - they're free to stick the switch into a threadlocal if they want that behaviour but they can also place it on a heap or run two side-by-side without them thrashing each other. we can of course keep adding obstacles and narrow the users choice to make more "convenient" to use but that rarely ends up satisfactory in the long run - when was the last time you created a local solution that is simple (like putting the counter in the pubsub instance which is a natural place for it to be anyway), then 6 months later went back and said, man I really should have used a global variable here - let's refactor it!